Add support for virtual network interfaces

jukkar opened the pull request:
With these commits, it is possible to stack network interfaces on top of each other. This can be used to implement IP tunneling and VPN support. There is a IPIP driver implementation included, it supports ipip, ipip6, ip6ip and ip6ip6 tunneling.

jukkar committed:
net: virtual: Add virtual network interface support

This can be used to implement tunneling, VPN etc. The virtual
interfaces can be chained together to support multilayer
network interfaces.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: virtual: Simple app for creating virtual interfaces

The app does not do much but shows how virtual network interfaces
support can be used.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: virtual: ipip: Add IP tunneling driver

Add a IP-to-IP tunneling driver. It supports both IPv4 and IPv6.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Allow app to disable IPv4 or IPv6 for the interface

Application can disable IPv4 or IPv6 later if those are not
needed nor used for a given network interface.

Fixes #14581

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print network interface flags values

Useful to know what flags are set for the network interface
so print them in the shell by the "net iface" command.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Mark IPv6 address valid for point2point links

As DAD is not done for point-to-point links, we can mark them
valid immediately. If this is not done, then the address will
never be picked as a source address to a sent packet.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: echo-server: Add tunneling interface support

Add support for IP tunneling with overlay-tunnel.conf file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Add function to set IPv4 ttl to interface

We had a functio to get IPv4 ttl but none for setting it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print network interface index instead of pointer

When printing network interface information print index or
index + pointer, so that the output is more user friendly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Add command to show virtual interface information

Add "net virtual" to show virtual network interface information.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: config: Try to only use auto started interface

If the first network interface is down when the net config init
is run, then the IP addresses etc would be set to wrong network
interface. So when initializing the network, try to first find
a network interface that is auto started and use that to configure
IP addresses etc.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: context: Print network interface index with pointer value

Makes debugging easier if interface index is printed with
the pointer.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: virtual: Add unit tests for checking tunneling

Add tests to verify that IP tunneling works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ipv4: Add utility func to create IPv4 header

Add a function that allow caller to create IPv4 header with
possibility to set various fields in the header.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ipv4: Discard pkt if TTL is 0

If the TTL value in the IPv4 header is 0, then the packet needs to
be discarded.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos commented on tests/net/virtual/Kconfig:
```diff
@@ -0,0 +1,26 @@
+# Private config options for echo-server sample app
+
+# Copyright (c) 2018 Intel Corporation
```
2021?

rlubos commented on tests/net/virtual/Kconfig:
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2018 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "Networking echo-server sample application"
```
This description doesn't seem right.

rlubos commented on subsys/net/l2/virtual/ipip/ipip.c:
```diff
@@ -0,0 +1,575 @@
+
+	NET_DBG("Starting iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
```
Should this comment be here? I'd expect it more in a sample code than an actual tunneling driver.

jukkar replied:
> Should this comment be here? I'd expect it more in a sample code than an actual tunneling driver.

Yeah, a left over from some other ideas. I will remove this.

rlubos commented on subsys/net/l2/virtual/ipip/ipip.c:
```diff
@@ -0,0 +1,575 @@
+
+	NET_DBG("Stopping iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
```
Same here.

rlubos reviewed (commented):
Pretty neat

jukkar commented on include/net/virtual.h:
```diff
@@ -0,0 +1,335 @@
+	 */
+	struct net_if *iface;
+
+	struct {
+		/** Carrier ON/OFF handler worker. This is used to create
+		 * network interface UP/DOWN event when ethernet L2 driver
+		 * notices carrier ON/OFF situation. We must not create another
+		 * network management event from inside management handler thus
+		 * we use worker thread to trigger the UP/DOWN event.
+		 */
+		struct k_work work;
+
+		/** Network interface that is detecting carrier ON/OFF event.
+		 */
+		struct net_if *iface;
+	} carrier_mgmt;
```
I think I will remove the carrier handling as that is handled by the real network interface.

jukkar commented on include/net/virtual_mgmt.h:
```diff
@@ -0,0 +1,135 @@
+#define _NET_VIRTUAL_INTERFACE_EVENT					\
+	(_NET_VIRTUAL_INTERFACE_BASE | NET_MGMT_EVENT_BIT)
+
+enum net_event_virtual_cmd {
+	NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_ON = 1,
+	NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_OFF,
+};
+
+#define NET_EVENT_VIRTUAL_INTERFACE_CARRIER_ON				\
+	(_NET_VIRTUAL_INTERFACE_EVENT |					\
+	 NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_ON)
+
+#define NET_EVENT_VIRTUAL_INTERFACE_CARRIER_OFF				\
+	(_NET_VIRTUAL_INTERFACE_EVENT |					\
+	 NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_OFF)
```
These would go away too.

jukkar commented on include/net/virtual_mgmt.h:
```diff
@@ -0,0 +1,135 @@
+
+/** @endcond */
+
+/**
+ * @brief Raise CARRIER_ON event when Virtual is connected.
+ *
+ * @param iface Virtual network interface.
+ */
+#if defined(CONFIG_NET_L2_VIRTUAL_MGMT)
+void virtual_interface_mgmt_raise_carrier_on_event(struct net_if *iface);
+#else
+static inline
+void virtual_interface_mgmt_raise_carrier_on_event(struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+}
+#endif
+
+/**
+ * @brief Raise CARRIER_OFF event when Virtual is disconnected.
+ *
+ * @param iface Virtual network interface.
+ */
+#if defined(CONFIG_NET_L2_VIRTUAL_MGMT)
+void virtual_interface_mgmt_raise_carrier_off_event(struct net_if *iface);
+#else
+static inline
+void virtual_mgmt_interface_raise_carrier_off_event(struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+}
+#endif
```
And these are not really needed.

jukkar commented on subsys/net/l2/virtual/virtual.c:
```diff
@@ -0,0 +1,448 @@
+
+	net_if_carrier_down(ctx->carrier_mgmt.iface);
+}
+
+static void handle_carrier(struct virtual_interface_context *ctx,
+			   struct net_if *iface,
+			   k_work_handler_t handler)
+{
+	k_work_init(&ctx->carrier_mgmt.work, handler);
+
+	ctx->carrier_mgmt.iface = iface;
+
+	k_work_submit(&ctx->carrier_mgmt.work);
+}
+
+int net_virtual_interface_carrier_on(struct net_if *iface)
+{
+	struct virtual_interface_context *ctx = net_if_l2_data(iface);
+
+	if (k_work_pending(&ctx->carrier_mgmt.work)) {
+		return -EAGAIN;
+	}
+
+	handle_carrier(ctx, iface, carrier_on);
+
+	return 0;
+}
+
+int net_virtual_interface_carrier_off(struct net_if *iface)
+{
+	struct virtual_interface_context *ctx = net_if_l2_data(iface);
+
+	if (k_work_pending(&ctx->carrier_mgmt.work)) {
+		return -EAGAIN;
+	}
+
+	handle_carrier(ctx, iface, carrier_off);
+
+	return 0;
+}
```
Would go away

jukkar commented on subsys/net/l2/virtual/virtual_mgmt.c:
```diff
@@ -0,0 +1,136 @@
+#include <net/net_if.h>
+#include <net/virtual_mgmt.h>
+
+void virtual_interface_mgmt_raise_carrier_on_event(struct net_if *iface)
+{
+	net_mgmt_event_notify(NET_EVENT_VIRTUAL_INTERFACE_CARRIER_ON, iface);
+}
+
+void virtual_interface_mgmt_raise_carrier_off_event(struct net_if *iface)
+{
+	net_mgmt_event_notify(NET_EVENT_VIRTUAL_INTERFACE_CARRIER_OFF, iface);
+}
```
Would be removed

rlubos reviewed (approved):
LGTM, awesome work 👍

jukkar commented:
Are there anyone else able to review?

mniestroj commented on subsys/net/ip/net_shell.c:
```diff
@@ -284,6 +284,53 @@ static void print_supported_ethernet_capabilities(
+				"IPv6,");
+	}
+
+	/* get rid of last ',' character */
+	str[pos - 1] = '\0';
```
What if there are no flags? Is that possible?

jukkar replied:
> What if there are no flags? Is that possible?

In theory that could be possible but in practice not. Anyway, I will add checks for the value.

mniestroj reviewed (approved):
I didn't get deep into virtual interface implementation, all the bits around seem fine. Just one comment, which is a possible out of buffer access.

nashif commented on tests/net/virtual/testcase.yaml:
```diff
@@ -0,0 +1,25 @@
+common:
+  depends_on: netif
+  min_ram: 32
+  tags: net tunnel virtual
```
please add integration_platforms and list the platforms where this should run on in CI for pull requests.

jukkar committed:
net: if: Add locking when setting/getting hoplimit or ttl

Locking was missing when setting or getting IPv6 hop limit
or IPv4 time-to-live values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
@nashif  integration_platforms added to tests, could you recheck

jukkar commented:
ping @nashif

jukkar commented:
ping @nashif
